### PR TITLE
Move app redirection logic to its own package

### PR DESCRIPTION
**Pull request by JoelSpeed**
## Description

This moves all of the code responsible for the application redirect into its own package.
Also rewrites the tests to use ginkgo.

## Motivation and Context

Trying to separate pieces of codes down to areas of concern.
This means that this package now contains all of this logic and we have an interface between this and the rest of the codebase.
This should allow us to write better tests for the main package in the future as we can ignore the specifics of the app redirect logic.

## How Has This Been Tested?

Unit tests only. No logical changes have been made.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: pkg/app/redirect/getters.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
+package redirect
```
Comment (NickMeves) [inline pkg/app/redirect/getters.go:1]: Getting these into their own file is a smart reorganization - I like it!

File: pkg/app/redirect/validator.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+// validator implements the Validator interface to allow validation
+// of redirect URLs.
+type validator struct {
+	allowedDomains []string
+}
+
+// IsValidRedirect checks whether the redirect URL is safe and allowed.
+func (v *validator) IsValidRedirect(redirect string) bool {
```
[the remark this answers is not in the document]
Comment (NickMeves) [reply pkg/app/redirect/validator.go:39]: While we have a PR open with arguably the ugliest method in our code base, are there any opportunities to implement this more clearly?

I think we have that scary warning from Github handled -- but this logic is so convoluted, I feel its easy for a bug in it to create an open redirect vulnerability.

File: pkg/requests/util/util.go [hunk 3]
```diff
@@ -52,3 +52,8 @@ func IsProxied(req *http.Request) bool {
 	}
 	return scope.ReverseProxy
 }
+
+func IsForwardedRequest(req *http.Request) bool {
```
Comment (NickMeves) [inline pkg/requests/util/util.go:56]: Good call, this is a much more sensible spot.

File: pkg/app/redirect/validator.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (NickMeves) [reply pkg/app/redirect/validator.go:39]: Or assuming this was a lift and shift -- we don't touch it 😅 

Pretty sure this monstrosity is the child of many CVE PRs.

[the remark this answers is not in the document]
Comment (JoelSpeed) [reply pkg/app/redirect/validator.go:39]: Yeah so this is all lift and shift, for now, to make it easier to review, I left it well alone! I can try and refactor it into a more understandable method but I'd argue that's scope creep and could/should be a separate PR

**Comment by JoelSpeed** [comment]
@NickMeves if you don't have any additional feedback, I think this is ready to merge

**Review by NickMeves — APPROVED** [review]
LGTM!

**Pull request merged**